File: tests/TestBody.cpp
#include "Motherboard.hpp"
#include "MotherboardAliases.hpp"
#include "Rcb4Adapter.hpp"
#include <cmath>

#include <chrono>
#include <thread>

int main()
{
  Roki::Motherboard mb;
  Roki::TTYConfig serviceConfig;

  serviceConfig.Port = "/dev/ttyAMA2";
  serviceConfig.Baudrate = 115200;
  serviceConfig.ParityBit = true;
  serviceConfig.Stopbits = serviceConfig.STOPBITS_ONE;
  serviceConfig.Timeout = 5;

  if (!mb.Configure(serviceConfig))
  {
    std::cout << mb.GetError() << std::endl;
    exit(1);
  }

  Roki::QueueInfo qi;

  Roki::Rcb4Adapter rcb4;

  rcb4.checkAcknowledge();

  bool ack = mb.BodySendSync(rcb4.GetRequestData(), rcb4.GetRequestSize(), rcb4.GetResponceData(), rcb4.GetResponceSize());

  if (!ack)
  {
    std::cout << mb.GetError() << std::endl;
    exit(1);
  }

  Rcb4BaseClass::ServoData sd;
  sd.Id = 8;
  sd.Sio = 1;

  int x = 0;
  while (true)
  {
    if (x < 500)
    {
      sd.Data = 7500 + int(500 * sin(0.1 * x));
      rcb4.setServoPos(&sd, 1, 5);
      if (!mb.BodySendAsync(rcb4.GetRequestData(), rcb4.GetRequestSize(), rcb4.GetResponceSize()))
      {
        std::cout << mb.GetError() << std::endl;
        exit(1);
      }
    }

    x += 1;

    if (x > 500)
    {
      rcb4.checkAcknowledge();
      ack = mb.BodySendSync(rcb4.GetRequestData(), rcb4.GetRequestSize(), rcb4.GetResponceData(), rcb4.GetResponceSize());
      if (!ack)
      {
        std::cout << mb.GetError() << std::endl;
        exit(1);
      }
    }

    if (!mb.GetQueueInfo(qi))
    {
      std::cout << mb.GetError() << std::endl;
      exit(1);
    }

    std::cout << "Body info: " << std::endl;
    std::cout << " Requests: " << qi.NumRequests << std::endl;
    std::cout << " Responces: " << qi.NumResponces << std::endl;
    std::cout << std::endl;

    if (qi.NumRequests == 0)
      break;

    // std::this_thread::sleep_for(std::chrono::seconds(1));
  }

  return 0;
}

File: examples/scripts/TestBody.py
import sys
import time

import Roki
from math import sin

mb = Roki.Motherboard()

conf = Roki.TTYConfig()

conf.Port = "/dev/ttyAMA2"
conf.Baudrate = 1250000
conf.Stopbits = Roki.Stopbits.One
conf.ParityBit = True
conf.Timeout = 2

mb.Configure(conf)
mb.SetQueuePeriod(5)

rcb = Roki.Rcb4(mb)

print("Connectiong to body...")
if(not rcb.checkAcknowledge()):
    print(rcb.getError())
    quit()

print("Ok")
#mb.ResetIMUCounter()

sd = Roki.Rcb4.ServoData()
sd.Id = 4
sd.Sio = 1

active = True

while active:
    x = 0
    req = 0
    while x < 2000:
        sd.Data = 7500 + int(500 * sin(0.1 * x))
        # Last argument is pause after command execution
        # Pause is measured in QueueTimer ticks
        # Pause = 0 means "Execute the command each tick"
        # 0 <= pause <= 255
        pause = 10
        if not rcb.setServoPosAsync([sd], int(5), pause):
            print(rcb.getError())
            active = False
            break
        req = mb.GetQueueInfo().NumRequests
        print(req)
        x = x + 1

input("Press any key to stop...")

mb.ResetQueue()
